src/Reflex/Class.hs

```diff
@@ -821,6 +825,63 @@ zipListWithEvent f l e = do
   mapAccumMaybe_ f' l e
 
 --------------------------------------------------------------------------------
+-- Sampleable
+--------------------------------------------------------------------------------
+
+-- | A 'Sampleable' type carries a function that can be applied to an 'Event' whenever the 'Event' fires.
+--
+-- 'Sampleable' is intended to be used via the '<@>' and '<@' operators and in conjunction with the `Applicative` interface.
+--
+-- If we have:
+--
+-- > f  :: a -> b -> c -> d
+-- > g  :: a -> b -> d
+-- > b1 :: Behavior t a
+-- > b2 :: Behavior t b
+-- > e  :: Event t c
+--
+-- and an 'Sampleable' instance for `Behavior`, then we get:
+--
+-- > f <$> b1 <*> b2 <@> e :: Event t d
+-- > g <$> b1 <*> b2 <@  e :: Event t d
+--
+-- It is a handy way to combine several 'Behavior's or 'Dynamic's together so that you can sample them all at the times when an `Event` is firing.
+--
+-- The 'Sampleable' class works via the 'attachApplyWithMaybeM' function so that we can implement a wider range of functions with good performance.
+-- Once you have the instance that you need, you should probably be using it via the '<@>' and '<@' operators.
+class Reflex t => Sampleable t f where
+  attachApplyWithMaybeM :: (a -> b -> PushM t (Maybe c)) -> f a -> Event t b -> Event t c
+
+instance Reflex t => Sampleable t Identity where
+  attachApplyWithMaybeM f (Identity a) = push $ \x -> f a x
+
+instance Reflex t => Sampleable t Maybe where
+  attachApplyWithMaybeM f (Just a) = push $ \x -> f a x
+  attachApplyWithMaybeM _ Nothing  = const never
```

Comment:
This instance strikes me as slightly dubious.  It relies heavily on the FunctorMaybe property of Events, whereas it seems that Sampleable may be susceptible to generalization beyond Events.  For example, imagine there were a type of Event with a fixed period (e.g. occurring once per second); it would not be in FunctorMaybe, but it would still totally make sense as something that could make use of a Sampleable.